### ISSUE TITLE
functionalSum.cpp and std::ranges::views::filter

### ISSUE THREAD

**Issue by ddalcino** [PacktPublishing/Hands-On-Functional-Programming-with-Cpp#2]
According to [cppreference](https://en.cppreference.com/w/cpp/compiler_support), gcc-10 now supports input range adaptors. I think that [std::ranges::filter_view](https://en.cppreference.com/w/cpp/ranges/filter_view) can be used to replace the filter function defined in functionalSum.cpp. This would also remove an unnecessary vector copy.

I know that gcc-10 wasn't available when this book was printed, and I understand if you want to maintain compatibility with older compilers.

**Comment by alexboly** [comment]
Thank you for your comment and for your interest in the code. 

I understand what you are saying, however the code from this repository has to stay in sync with the code included in the book. Due to the C++ standard's situation when the book was published, the ranges library is treated separately in Chapter 14. Indeed, the compiler support when the book was published was limited for the ranges library.  

In addition, for learning purpose I believe it's best to start from very simple code constructs that showcase the fundamental elements of functional programming, to help demystifying it. I repeatedly made the point in the book that the code presented is meant to show concepts and is not necessarily optimal in terms of memory or CPU consumption - until chapter 10 which is fully dedicated on performance optimization.

In other words, I made a deliberate choice to trust the readers' ability to focus on the topic at hand and temporarily ignore optimization. I believe the book has a better learning flow in this way - but I'm willing to learn from the readers' feedback :).

**Comment by ddalcino** [comment]
Ok, that's fair.
